### PR TITLE
FEAT Adding Python example for unit test with mock (pattern 05)

### PR DESCRIPTION
**Pull request by wtromano-aws**
*Issue #, if available:*

Issue 47

*Description of changes:*

Adds an example of local testing of Lambda using mocks for test pattern 05.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python-test-samples/lambda-mock/README.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+  - [Make commands for test and deployment](#make-commands-for-test-and-deployment)
+  - [Running the project](#running-the-project)
+  - [License](#license)
+---
+
+## System Under Test (SUT)
+
+The SUT in this pattern is a Lambda function that makes calls to other AWS cloud services using an AWS SDK.
+
+![System Under Test](/_img/pattern_04_lambda_mock_sut.png)
```
Comment (dancfox) [inline python-test-samples/lambda-mock/README.md:33]: Incorrect diagram? This is the same diagram as below. Tests and synthetic event are not part of the SUT.

File: python-test-samples/lambda-mock/README.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+* ```make deploy```: Deploy the stack to an AWS Account
+* ```make deploy.g```: Deploy the stack to an AWS Account, prompting for stack parameters
+
+[Top](#contents)
+
+---
+
+## Running the project
+
+* First, create the Python Virtual Environment by using the ```make createEnv``` command.
```
Comment (dancfox) [inline python-test-samples/lambda-mock/README.md:111]: wrong make command?

File: python-test-samples/lambda-mock/README.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+* Run the unit tests with the ```make unittest``` command.
```
Comment (dancfox) [inline python-test-samples/lambda-mock/README.md:112]: wrong command

File: python-test-samples/lambda-mock/README.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+* To deploy the project, deploy the project using the ```make deploy.guided``` command.
```
Comment (dancfox) [inline python-test-samples/lambda-mock/README.md:113]: wrong command

File: python-test-samples/lambda-mock/README.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  * ```{ "PK" : "C#TestCustomer", "data" : "Testing Customer"}```
+  * ```{ "PK" : "D#Welcome", "data" : "\nHello and Welcome!\n"}```
+* As authentication is not configured, you can test the endpoint using the API Gateway service in the AWs Console:
+*  ```https://{region}.console.aws.amazon.com/apigateway```  
+
+[Top](#contents)
+
+---
+
+## License
```
Comment (dancfox) [inline python-test-samples/lambda-mock/README.md:124]: License not necessary here.

**Review by dancfox — CHANGES_REQUESTED** [review]
Please see comments! I will do another round once you attend to a few fixes. Thank you!

File: python-test-samples/lambda-mock/LICENSE [hunk 6]
```diff
@@ -0,0 +1,15 @@
+Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (wtromano-aws) [reply python-test-samples/lambda-mock/LICENSE]: Removed, thank you!

File: python-test-samples/lambda-mock/Makefile [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+#  Helpers  #
+#############
+
+define HELP_MESSAGE
+
+	Usage: make <command>
+
+	Commands:
+
+	build     Build Lambda function dependencies
```
[the remark this answers is not in the document]
Comment (wtromano-aws) [reply python-test-samples/lambda-mock/Makefile]: Corrected, great catch.

File: python-test-samples/lambda-mock/README.md (hunk 2 above)
Comment (wtromano-aws) [reply python-test-samples/lambda-mock/README.md]: I missed a whole section of the README when I re-tooled the Makefile, thank you for finding it.   Resolved.

File: python-test-samples/lambda-mock/README.md (hunk 4 above)
Comment (wtromano-aws) [reply python-test-samples/lambda-mock/README.md]: same.

File: python-test-samples/lambda-mock/Makefile [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+	deploy.g  Deploy guided (for initial deployment)
+	deploy    Deploy subsequent changes
+
+	install   Install Pipenv, application and dev dependencies defined in Pipfile
+	shell     Spawn a virtual environment shell
+	deps      Install project dependencies locally
+	test      Run unit test locally using mockets
+	coverage  Run unit test coverage reports
+	scan	  Run cose scanning tools
```
[the remark this answers is not in the document]
Comment (wtromano-aws) [reply python-test-samples/lambda-mock/Makefile]: corrected, thank you.

File: python-test-samples/lambda-mock/CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,59 @@
+# Contributing Guidelines
```
[the remark this answers is not in the document]
Comment (wtromano-aws) [reply python-test-samples/lambda-mock/CONTRIBUTING.md]: removed.

File: python-test-samples/lambda-mock/Makefile [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+	Usage: make <command>
+
+	Commands:
+
+	build     Build Lambda function dependencies
+	deploy.g  Deploy guided (for initial deployment)
+	deploy    Deploy subsequent changes
+
+	install   Install Pipenv, application and dev dependencies defined in Pipfile
```
[the remark this answers is not in the document]
Comment (wtromano-aws) [reply python-test-samples/lambda-mock/Makefile]: copy-paste issue!  resolved in the latest push.

File: python-test-samples/lambda-mock/Makefile [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+	shell     Spawn a virtual environment shell
+	deps      Install project dependencies locally
+	test      Run unit test locally using mockets
```
[the remark this answers is not in the document]
Comment (wtromano-aws) [reply python-test-samples/lambda-mock/Makefile]: mockets = mocked muppets?
Great catch, thank you.  Resolved.

File: python-test-samples/lambda-mock/CONTRIBUTING.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (wtromano-aws) [reply python-test-samples/lambda-mock/CONTRIBUTING.md]: removed

File: python-test-samples/lambda-mock/README.md (hunk 1 above)
Comment (wtromano-aws) [reply python-test-samples/lambda-mock/README.md:32]: I triple-checked, and this is the diagram from our inventory.  Perhaps the inventory is also incorrect?

Comment (dancfox) [reply python-test-samples/lambda-mock/README.md:32]: Okay you are right. The inventory doc is wrong. However, this image does not accurately represent the architecture of the system under test in your sample. Could you create a new image from the PPT that does? 

Representations of tests should not appear in SUT diagrams.

**Comment by dancfox** [comment]
Ran into the following error.

Screenshot 2023-03-03 at 2 50 50 PM

**Comment by wtromano-aws** [comment]
Thank you Dan, I updated the diagrams in the inventory and this repo.
Also, I think I addressed the issue in the Makefile, let me know how the edit works for you.

**Review by dancfox — APPROVED** [review]

**Pull request merged**